fastly/block_fastly_service_v1_gzip.go

```diff
@@ -173,6 +175,33 @@ func (h *GzipServiceAttributeHandler) Read(d *schema.ResourceData, s *gofastly.S
 
 	gl := flattenGzips(gzipsList)
 
+	// Note: Although "content_types" and "extensions" fields are optional in spec,
+	// Fastly API will actually set the default value silently when these fields are not sent
+	// or an empty field value is sent. This will cause unexpected diff.
+	// We need to ignore these fields in the API response unless field values are explicitly set.
+	type IgnoreFields struct {
+		Name string
+	}
+	ignoreList := map[string][]IgnoreFields{}
+
+	for _, elem := range d.Get("gzip").(*schema.Set).List() {
+		name := elem.(map[string]interface{})["name"].(string)
+		if elem.(map[string]interface{})["content_types"].(*schema.Set).Len() == 0 {
+			ignoreList[name] = append(ignoreList[name], IgnoreFields{Name: "content_types"})
+		}
+		if elem.(map[string]interface{})["extensions"].(*schema.Set).Len() == 0 {
+			ignoreList[name] = append(ignoreList[name], IgnoreFields{Name: "extensions"})
+		}
+	}
+
+	for i, g := range gl {
+		if v, ok := ignoreList[g["name"].(string)]; ok && len(v) > 0 {
+			for _, sl := range v {
+				gl[i][sl.Name] = nil
+			}
+		}
+	}
```

Comment:
This might be a good use case for an extra block scope as a delimiter for the specialised behaviour...

```go
// NOTE: Although "content_types" and "extensions" fields are optional in spec...etc
{
	type IgnoreFields struct {
		Name string
	}
	// ...and the rest of the new code here.
}
```

This is a nice _visual_ indicator of the boundary of the new code, and in future if we want to remove the code it's easier to identify where the boundary is for the specialised logic.